client/grpc_query.go

```diff
@@ -72,39 +68,51 @@ func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, req, reply i
 		ctx = ctx.WithHeight(height)
 	}
 
-	abciReq := abci.RequestQuery{
-		Path:   method,
-		Data:   reqBz,
-		Height: ctx.Height,
-	}
+	if ctx.GRPC != nil {
+		err = ctx.GRPC.Invoke(grpcCtx, method, req, reply, opts...)
+		if err != nil {
+			return err
+		}
+	} else {
+		reqBz, err := protoCodec.Marshal(req)
+		if err != nil {
+			return err
+		}
 
-	res, err := ctx.QueryABCI(abciReq)
-	if err != nil {
-		return err
-	}
+		abciReq := abci.RequestQuery{
+			Path:   method,
+			Data:   reqBz,
+			Height: ctx.Height,
+		}
 
-	err = protoCodec.Unmarshal(res.Value, reply)
-	if err != nil {
-		return err
-	}
+		res, err := ctx.QueryABCI(abciReq)
+		if err != nil {
+			return err
+		}
 
-	// Create header metadata. For now the headers contain:
-	// - block height
-	// We then parse all the call options, if the call option is a
-	// HeaderCallOption, then we manually set the value of that header to the
-	// metadata.
-	md = metadata.Pairs(grpctypes.GRPCBlockHeightHeader, strconv.FormatInt(res.Height, 10))
-	for _, callOpt := range opts {
-		header, ok := callOpt.(grpc.HeaderCallOption)
-		if !ok {
-			continue
+		err = protoCodec.Unmarshal(res.Value, reply)
+		if err != nil {
+			return err
 		}
 
-		*header.HeaderAddr = md
-	}
+		// Create header metadata. For now the headers contain:
+		// - block height
+		// We then parse all the call options, if the call option is a
+		// HeaderCallOption, then we manually set the value of that header to the
+		// metadata.
+		md = metadata.Pairs(grpctypes.GRPCBlockHeightHeader, strconv.FormatInt(res.Height, 10))
+		for _, callOpt := range opts {
+			header, ok := callOpt.(grpc.HeaderCallOption)
+			if !ok {
+				continue
+			}
+
+			*header.HeaderAddr = md
+		}
 
-	if ctx.InterfaceRegistry != nil {
-		return types.UnpackInterfaces(reply, ctx.InterfaceRegistry)
+		if ctx.InterfaceRegistry != nil {
+			return types.UnpackInterfaces(reply, ctx.InterfaceRegistry)
+		}
```

Comment:
We should also call this code in the case where `ctx.GRPC != nil`, no?

Reply:
I'm not sure actually, good call